handlers/handlers.go

```diff
@@ -1,238 +1,238 @@
+	} 
+} 
+ 
+func HandleUserRequest( 
+	responseWriter http.ResponseWriter, 
+	request *http.Request, 
+) { 
+	type SignupForm struct { 
+		DisplayName  string `json:"displayName"` 
+		EmailAddress string `json:"emailAddress"` 
+		Password     string `json:"password"` 
+	} 
+ 
+	switch request.Method { 
+	case http.MethodPost: 
+		var signupForm SignupForm 
+		body := getRequestBody(request) 
+ 
+		if err := json.Unmarshal(body, &signupForm); err != nil { 
+			panic(err) 
+		} 
+ 
+		var statusCode int 
+		if err := userservice.StoreNewUser( 
+			signupForm.DisplayName, 
+			signupForm.EmailAddress, 
+			signupForm.Password, 
+		); err != nil { 
+			if err == userservice.EmailAddressAlreadyInUseError { 
+				statusCode = http.StatusConflict 
+			} else { 
+				panic(err) 
+			} 
+		} else { 
+			statusCode = http.StatusCreated 
+		} 
+ 
+		responseWriter.WriteHeader(statusCode) 
+ 
+	default: 
+		respondWithMethodNotAllowed(responseWriter, []string{http.MethodPost}) 
+	} 
+} 
+ 
+func HandleSessionRequest( 
+	responseWriter http.ResponseWriter, 
+	request *http.Request, 
+) { 
+	type LoginForm struct { 
+		EmailAddress string `json:"emailAddress"` 
+		Password     string `json:"password"` 
+	} 
+ 
+	switch request.Method { 
+	case http.MethodPost: 
+		var loginForm LoginForm 
+		body := getRequestBody(request) 
+ 
+		if err := json.Unmarshal(body, &loginForm); err != nil { 
+			panic(err) 
+		} 
+ 
+		if err := userservice.AuthenticateUserCredentials( 
+			loginForm.EmailAddress, 
+			loginForm.Password, 
+		); err != nil { 
+			panic(err) 
+		} 
+ 
+		// Set our cookie to have a valid JWT Token as the value 
+		{ 
+			userId, err := userservice.GetIdForUserWithEmailAddress(loginForm.EmailAddress) 
+			if err != nil { 
+				panic(err) 
+			} 
+ 
+			token, err := createTokenAsString(userId, credentialTimeoutInMinutes) 
+			if err != nil { 
+				panic(err) 
+			} 
+ 
+			expirationTime := time.Now().Add(credentialTimeoutInMinutes * time.Minute) 
+ 
+			cookie := http.Cookie{ 
+				Name:     cerealNotesCookieName, 
+				Value:    token, 
+				Expires:  expirationTime, 
+				HttpOnly: true, 
+			} 
+ 
+			http.SetCookie(responseWriter, &cookie) 
+		} 
+ 
+		responseWriter.WriteHeader(http.StatusCreated) 
+		responseWriter.Write([]byte(fmt.Sprint("passward email combo was correct"))) 
+ 
+	case http.MethodDelete: 
+		// Cookie will overwrite existing cookie then delete itself 
+		cookie := http.Cookie{ 
+			Name:     cerealNotesCookieName, 
+			Value:    "", 
+			HttpOnly: true, 
+			MaxAge:   -1, 
+		} 
+ 
+		http.SetCookie(responseWriter, &cookie) 
+		responseWriter.WriteHeader(http.StatusOK) 
+		responseWriter.Write([]byte(fmt.Sprint("user succefully logged out"))) 
+	default: 
+		respondWithMethodNotAllowed( 
+			responseWriter, 
+			[]string{http.MethodPost, http.MethodDelete}) 
+	} 
+} 
+ 
+type AuthentictedRequestHandlerType func( 
+	http.ResponseWriter, 
+	*http.Request, 
+	models.UserId) 
+ 
+func AuthenticateOrRedirectToLogin( 
```

Comment:
omg I love this function so much. A function that takes a function and returns another function. This smells so much of haskell

Reply:
I think I might adopt this pattern for authenticated endpoints at work

Reply:
Minor: The return value of this function is 
`func(http.ResponseWriter, *http.Request)`
you could change it to
`http.HandlerFunc`, which is the same thing, but is more concise and possibly more readable (there's an argument to be made that it's no more readable. So ... whatevs)

Reply:
error: handlers/handlers.go:202:3: http.HandleFunc is not a type

Reply:
HandlerFunc, not HandleFunc